core/shared/src/main/scala/scalaz/zio/ZIO.scala

```diff
@@ -286,11 +286,17 @@ sealed trait ZIO[-R, +E, +A] extends Serializable { self =>
     tryOrElse(that.map(Right(_)), ZIO.succeedLeft)
 
   private final def tryOrElse[R1 <: R, E2, B](that: => ZIO[R1, E2, B], succ: A => ZIO[R1, E2, B]): ZIO[R1, E2, B] =
-    (self.tag: @switch) match {
-      case ZIO.Tags.Fail => that
-
-      case _ => new ZIO.Fold[R1, E, E2, A, B](self, _ => that, succ)
-    }
+    new ZIO.Fold[R1, E, E2, A, B](
+      self,
+      cause =>
+        if (!cause.died) that
+        else
+          cause.stripFailures match {
+            case Some(c) => ZIO.halt(c)
+            case None    => ZIO.dieMessage("Logic error in tryOrElse: stripFailures returned None")
```

Comment:
We could eliminate this by changing `stripFailures` to return `Either[List[E], Cause[Nothing]]`. In which case maybe it should be called `filterFailures`. Then you match agains the either and either `that` or  `ZIO.halt(_)` 

I'll leave it up to you though!